## Prepare dataset in the required format

### image url

The uzh image urls are not publicly accessible. The images paths on google colab are not https public urls. So upload the images to github and then re-extract the public https urls.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!rm -rf /content/uzh_artefacts_images # already cloned it before, delete the folder first

In [ ]:
# copy images from drive to content(Colab’s temporary runtime environment), because Git can only push files that are in the Colab environment (not on Drive).
import shutil
import os

src_folder = "/content/drive/MyDrive/master_thesis_project/uzh_images"
dst_folder = "/content/uzh_artefacts_images"
os.makedirs(dst_folder, exist_ok=True)

for f in os.listdir(src_folder):
    if f.lower().endswith(".jpg"):
        shutil.copy(os.path.join(src_folder, f), dst_folder)

In [ ]:
%cd /content/uzh_artefacts_images

# Initialize git repo
!git init

# Set Git identity
!git config user.name "juanitalfr5-pixel"
!git config user.email "juanita.lfr5@gmail.com"

# Stage and commit files
!git add .
!git commit -m "uploading UZH artefact images"

# Make sure branch name matches GitHub’s default
!git branch -M main

# Remove any previous remote (in case it exists)
!git remote remove origin || true

# Add remote with PAT
!git remote add origin https://xxx/uzh_artefacts_images.git

# Push to GitHub
!git push -u origin main

/content/uzh_artefacts_images
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint: 
hint: 	git config --global init.defaultBranch <name>
hint: 
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint: 
hint: 	git branch -m <name>
Initialized empty Git repository in /content/uzh_artefacts_images/.git/
[master (root-commit) fbdad4e] uploading UZH artefact images
 4675 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 1510_image_0.jpg
 create mode 100644 1510_image_1.jpg
 create mode 100644 1510_image_10.jpg
 create mode 100644 1510_image_11.jpg
 create mode 100644 1510_image_12.jpg
 create mode 100644 1510_image_13.jpg
 create mode 100644 1510_image_14.jpg
 create mode 100644 1510_image_15.j

Git counts hidden files too.  
When initializing the repo, Git created two hidden items automatically:  
.gitignore (maybe, depending on settings).
.gitattributes or .git folder internals.
Those can account for the 2 extra objects — so 4675 images + 2 metadata files = 4677 total objects.

In [ ]:
!git ls-files | grep -i '.jpg' | wc -l

4675


generate https publicly accessible image paths

In [ ]:
import os

repo_base_url = "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main"
repo_folder = "/content/uzh_artefacts_images"

image_urls = []

for f in os.listdir(repo_folder):
    if f.lower().endswith(".jpg"):
        image_urls.append(f"{repo_base_url}/{f}")

In [ ]:
image_urls[:5]  # show first 5 URLs

['https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/64974_image_4.jpg',
 'https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/5708_image_5.jpg',
 'https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7843_image_0.jpg',
 'https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/5468_image_0.jpg',
 'https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6534_image_7.jpg']

https://techcommunity.microsoft.com/blog/azure-ai-foundry-blog/announcing-new-fine-tuning-capabilities-with-images-on-azure-openai-service/4303695?utm_source=chatgpt.com

Image Dataset Requirements.

To ensure the best performance and compliance, there are specific requirements for your image datasets:  

Size: Your training file can contain up to 50,000 examples with images, with each example having a maximum of 64 images. Each image can be up to 10 MB.  
Format: Images must be in JPEG, PNG, or WEBP format and in RGB or RGBA mode. Images cannot be included as output from messages with the assistant role.  
Content Moderation: Images are scanned before training to ensure compliance with our usage policy. Images containing people, faces, or CAPTCHAs will be excluded from the dataset.  

Handling Skipped Images  

If your images are skipped during the training process, it could be due to several reasons such as containing CAPTCHAs, people, faces, inaccessible URLs, large file sizes, invalid mode or invalid formats. Ensure your images meet the specified requirements to avoid these issues.  

Uploading Large Files  

For large training files, you can upload files up to 8 GB in multiple parts using the Uploads API. This is particularly useful for extensive datasets that exceed the 512 MB limit of the Files API.  

Reducing Training Costs  

To optimize training costs, you can set the detail parameter for an image to low, which resizes the image to 512 by 512 pixels and represents it by 85 tokens regardless of its size. This reduces the cost of training while maintaining the quality of the model.  

Additional Considerations  

To control the fidelity of image understanding, you can set the detail parameter of image_url to low, high, or auto for each image. This affects the number of tokens per image that the model sees during training and impacts the cost of training.  

We are thrilled to see how you will leverage these new capabilities to create innovative and engaging AI applications. For more detailed information, please refer to our documentation on Azure OpenAI Service.  

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# check whether there are artefacts that have more than 64 images
import os

img_folder = "/content/drive/My Drive/master_thesis_project/uzh_images"
supported_formats = (".jpg", ".jpeg", ".png", ".webp")

# List all files in the folder
all_files = os.listdir(img_folder)
print(f"Total files in folder: {len(all_files)}\n")

# Filter image files
image_files = [f for f in all_files if f.lower().endswith(supported_formats)]
print(f"Total supported image files: {len(image_files)}\n")

# Check for unsupported formats
for f in all_files:
    if not f.lower().endswith(supported_formats):
        print(f"Unsupported format: {f}")

# Check for very large number of images per artifact
from collections import defaultdict

# Assuming your naming scheme is id_image_idx.jpg
artifact_counts = defaultdict(int)
for f in image_files:
    artifact_id = f.split("_")[0]
    artifact_counts[artifact_id] += 1

for artifact_id, count in artifact_counts.items():
    if count > 10:
        print(f"Artifact {artifact_id} has {count} images (consider reducing to 10)")

Total files in folder: 4675

Total supported image files: 4675

Artifact 1510 has 67 images (consider reducing to 64)
Artifact 7563 has 65 images (consider reducing to 64)
Artifact 5360 has 65 images (consider reducing to 64)


replace the image paths in the json file with the github image paths and make sure maximal 64 images for each artefact. use multi-turn chat file format.  

https://learn.microsoft.com/en-us/azure/ai-foundry/openai/how-to/fine-tuning?tabs=azure-openai&pivots=rest-api

All sources that write 'max 64 images' are from the microsoft platform, but on the openai platform https://platform.openai.com/docs/guides/vision-fine-tuning#size, it writes 'max 10 images per sample'. I decide to follow the openai platform.


optimize the promt(Write effective prompts):  
before:  "content": "Du bist ein hilfreicher Assistent, der Metadaten und Beschreibungen für Artefakte auf Deutsch generiert."  
after:  "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Metadaten und Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."

In [ ]:
import json
import re
import os

input_file = "/content/drive/My Drive/master_thesis_project/uzh_metadata_compact.json"
output_file = "/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only.jsonl"

# GitHub raw URL prefix
github_prefix = "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/"

with open(input_file, "r", encoding="utf-8") as f:
    data = json.load(f)

with open(output_file, "w", encoding="utf-8") as f_out:

    for obj in data:
        beschreibung_raw = obj.get("description", "")
        beschreibung = re.split(r"(?=Nach:|Zur Kontextualisierung)", beschreibung_raw)[0].strip() # Remove parts starting with "Nach:" or "Zur Kontextualisierung"

        images = obj.get("images", [])
        if not images:
            continue

        # Replace local paths with GitHub URLs
        github_images = []
        for img_path in images[:10]: # ensure max 10 images
            filename = os.path.basename(img_path)  # extract e.g., "4121_image_0.jpg"
            github_images.append({"type": "image_url", "image_url": {"url": github_prefix + filename}})

        # build conversation
        messages = [
            {
                "role": "system",
                "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."
            },
            {
                "role": "user",
                "content": github_images + [
                    {"type": "text", "text": "Beschreibe dieses Artefakt detailliert inklusive Form, Dekoration, Farben und Erhaltungszustand."}
                ]
            },
            {
                "role": "assistant",
                "content": f"{beschreibung if beschreibung else 'Keine Beschreibung im Datensatz vorhanden.'}"
            }
        ]

        entry = {"messages": messages}
        f_out.write(json.dumps(entry, ensure_ascii=False) + "\n")

print(f"\nDone! Single-turn JSONL with GitHub image URLs written to {output_file}")


Done! Single-turn JSONL with GitHub image URLs written to /content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only.jsonl


In [ ]:
# Show the first 5 lines of the JSONL file
with open(output_file, "r", encoding="utf-8") as f:
    for i, line in enumerate(f):
        if i >= 5:
            break
        print(line.strip())

{"messages": [{"role": "system", "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."}, {"role": "user", "content": [{"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_0.jpg"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_1.jpg"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_2.jpg"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_3.jpg"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_4.jpg"}}, {"type": 

In [ ]:
import json

# Show one pretty-printed sample from your JSONL file
with open(output_file, "r", encoding="utf-8") as f:
    for i, line in enumerate(f):
        if i == 0:  # just take the first artefact
            obj = json.loads(line)
            print(json.dumps(obj, indent=2, ensure_ascii=False))
            break

{
  "messages": [
    {
      "role": "system",
      "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."
    },
    {
      "role": "user",
      "content": [
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_0.jpg"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_1.jpg"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_2.jpg"
          }
        },
        {
          "type": "image_url",
          "image_url": {

In [ ]:
# Analyze the current description data
import json

# Initialize counters
description_lengths = []
char_counts = []
word_counts = []

with open(output_file, "r", encoding="utf-8") as f:
    for line_num, line in enumerate(f, 1):
        try:
            data = json.loads(line)
            # Get the assistant's content (description)
            assistant_content = data['messages'][2]['content']

            char_count = len(assistant_content)
            word_count = len(assistant_content.split())

            char_counts.append(char_count)
            word_counts.append(word_count)
            description_lengths.append((line_num, char_count, word_count, assistant_content[:100] + "..."))

        except (json.JSONDecodeError, KeyError, IndexError) as e:
            print(f"Error parsing line {line_num}: {e}")

# Print statistics
print("=== DESCRIPTION LENGTH ANALYSIS ===")
print(f"Total examples: {len(char_counts)}")
print(f"Average characters: {sum(char_counts)/len(char_counts):.1f}")
print(f"Average words: {sum(word_counts)/len(word_counts):.1f}")
print(f"Min characters: {min(char_counts)}")
print(f"Max characters: {max(char_counts)}")
print(f"Min words: {min(word_counts)}")
print(f"Max words: {max(word_counts)}")
print()

# Count by quality tiers
short_count = sum(1 for wc in word_counts if wc < 50)
medium_count = sum(1 for wc in word_counts if 50 <= wc < 100)
medium_long_count = sum(1 for wc in word_counts if 100 <= wc < 150)  # This variable exists
long_count = sum(1 for wc in word_counts if wc >= 150)

print("=== QUALITY DISTRIBUTION ===")
print(f"Poor (<50 words): {short_count} examples ({short_count/len(word_counts)*100:.1f}%)")
print(f"Medium (50-99 words): {medium_count} examples ({medium_count/len(word_counts)*100:.1f}%)")
print(f"Medium-long (100-149 words): {medium_long_count} examples ({medium_long_count/len(word_counts)*100:.1f}%)")  # FIXED
print(f"Good (150+ words): {long_count} examples ({long_count/len(word_counts)*100:.1f}%)")
print()

# Show extremes
print("=== SHORTEST DESCRIPTIONS (Potential problems) ===")
shortest = sorted(description_lengths, key=lambda x: x[2])[:10]
for line_num, chars, words, preview in shortest:
    print(f"Line {line_num}: {words} words, {chars} chars - {preview}")

print("\n=== LONGEST DESCRIPTIONS (Good examples) ===")
longest = sorted(description_lengths, key=lambda x: x[2], reverse=True)[:10]
for line_num, chars, words, preview in longest:
    print(f"Line {line_num}: {words} words, {chars} chars - {preview}")

=== DESCRIPTION LENGTH ANALYSIS ===
Total examples: 495
Average characters: 691.3
Average words: 94.3
Min characters: 105
Max characters: 9264
Min words: 12
Max words: 1278

=== QUALITY DISTRIBUTION ===
Poor (<50 words): 210 examples (42.4%)
Medium (50-99 words): 147 examples (29.7%)
Medium-long (100-149 words): 67 examples (13.5%)
Good (150+ words): 71 examples (14.3%)

=== SHORTEST DESCRIPTIONS (Potential problems) ===
Line 28: 12 words, 105 chars - Vorderer, abgesägter Teil eines Stirnziegel (Antefix) mit Gorgoneion.

Hellgrüner, leicht gemagerter...
Line 56: 15 words, 124 chars - Die Fussunterseite ist tongrundig.

Feiner gelbbrauner hart gebrannter Ton mit mattglänzendem bräunl...
Line 105: 15 words, 122 chars - Die Fussunterseite ist tongrundig.

Feiner gelblicher hart gebrannter Ton mit glänzendem braunschwar...
Line 120: 15 words, 115 chars - Die Fussunterseite ist tongrundig.

Feiner graugelber hart gebrannter Ton mit mattem graubraunem Fir...
Line 126: 15 words, 123 chars - D

Large images would be more tokens and cost more. According to https://platform.openai.com/docs/guides/vision-fine-tuning#size :
If you set the detail parameter for an image to low, the image is resized to 512 by 512 pixels and is only represented by 85 tokens regardless of its size. This will reduce the cost of training.

In [ ]:
import os
from PIL import Image

# Path to your image folder
image_folder = "/content/drive/MyDrive/master_thesis_project/uzh_images"

# List all files in that folder
image_files = [f for f in os.listdir(image_folder) if f.lower().endswith('.jpg')]

too_large = []

for file_name in image_files:
    image_path = os.path.join(image_folder, file_name)
    try:
        with Image.open(image_path) as img:
            width, height = img.size
            if width > 512 or height > 512:
                too_large.append((file_name, width, height))
    except Exception as e:
        print(f"Could not open {file_name}: {e}")

# Show results
if too_large:
    print(f"{len(too_large)} out of {len(image_files)} images are larger than 512px:")
    for name, w, h in too_large:
        print(f" - {name}: {w}x{h}px")
else:
    print("All images are 512px or smaller. Perfect!")

4675 out of 4675 images are larger than 512px:
 - 2240_image_3.jpg: 914x1200px
 - 2240_image_4.jpg: 914x1200px
 - 2240_image_5.jpg: 914x1200px
 - 2240_image_6.jpg: 1200x951px
 - 2240_image_7.jpg: 1200x964px
 - 2240_image_8.jpg: 1200x1193px
 - 2240_image_9.jpg: 1200x1051px
 - 2240_image_10.jpg: 1200x1051px
 - 2240_image_11.jpg: 1200x916px
 - 2240_image_12.jpg: 1200x904px
 - 2240_image_13.jpg: 778x1200px
 - 2240_image_14.jpg: 949x1200px
 - 2240_image_15.jpg: 778x1200px
 - 2240_image_16.jpg: 802x1200px
 - 2240_image_17.jpg: 778x1200px
 - 2240_image_18.jpg: 778x1200px
 - 2240_image_19.jpg: 778x1200px
 - 2240_image_20.jpg: 778x1200px
 - 2240_image_21.jpg: 778x1200px
 - 2240_image_22.jpg: 1108x1200px
 - 2240_image_23.jpg: 1108x1200px
 - 2240_image_24.jpg: 718x1200px
 - 2240_image_25.jpg: 718x1200px
 - 2240_image_26.jpg: 927x1200px
 - 2240_image_27.jpg: 940x1200px
 - 2240_image_28.jpg: 1034x1200px
 - 2240_image_29.jpg: 1043x1200px
 - 2240_image_30.jpg: 1160x1200px
 - 2240_image_31.jpg: 1143x1

In [ ]:
import os
import json

jsonl_file = "/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only.jsonl"

# Check file size
file_size_mb = os.path.getsize(jsonl_file) / (1024 * 1024)
print(f"File size: {file_size_mb:.2f} MB")

# Check number of images per example
max_images_allowed = 10
supported_formats = (".jpg", ".jpeg", ".png", ".webp")
max_10_images = True

with open(jsonl_file, "r", encoding="utf-8") as f:
    for i, line in enumerate(f):
        try:
            example = json.loads(line)
            user_content = example["messages"][1]["content"]
            if not isinstance(user_content, list):
                print(f"Example {i}: user content is not a list")
                continue

            # Extract image URLs
            image_urls = [x["image_url"]["url"] for x in user_content if x.get("type") == "image_url"]

            if len(image_urls) > max_images_allowed:
                print(f"Example {i}: has {len(image_urls)} images (consider reducing to {max_images_allowed})")
                max_10_images = False

            # check formats
            for url in image_urls:
                if not url.lower().endswith(supported_formats):
                    print(f"Example {i}: unsupported image format -> {url}")

        except Exception as e:
            print(f"Error processing example {i}: {e}")

if max_10_images:
    print(f"All the artefacts have maximal 10 images.")

File size: 1.04 MB
All the artefacts have maximal 10 images.


https://community.openai.com/t/gpt-4-vision-preview-fidelity-detail-parameter/477563

add the parameter detail=low into data  

e.g. taken from https://platform.openai.com/docs/guides/vision-fine-tuning#size  

{
  "type": "image_url",
  "image_url": {
    "url": "https://upload.wikimedia.org/wikipedia/commons/3/36/Danbo_Cheese.jpg",
    "detail": "low"
  }
}

In [ ]:
import json

# Input and output paths
input_file = "/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only.jsonl"
output_file = "/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only_detail_low.jsonl"

with open(input_file, "r", encoding="utf-8") as infile, open(output_file, "w", encoding="utf-8") as outfile:
    for line in infile:
        data = json.loads(line)

        # Go through messages
        for message in data.get("messages", []):
            if isinstance(message.get("content"), list):
                for item in message["content"]:
                    if (isinstance(item, dict) and
                        item.get("type") == "image_url" and
                        "image_url" in item):
                        item["image_url"]["detail"] = "low"

        # Write updated data
        outfile.write(json.dumps(data, ensure_ascii=False) + "\n")

print("All image entries updated with detail='low' and saved to:")
print(output_file)

All image entries updated with detail='low' and saved to:
/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only_detail_low.jsonl


In [ ]:
import json

# Show one pretty-printed sample from your JSONL file
with open(output_file, "r", encoding="utf-8") as f:
    for i, line in enumerate(f):
        if i == 0:  # just take the first artefact
            obj = json.loads(line)
            print(json.dumps(obj, indent=2, ensure_ascii=False))
            break

{
  "messages": [
    {
      "role": "system",
      "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."
    },
    {
      "role": "user",
      "content": [
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_0.jpg",
            "detail": "low"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_1.jpg",
            "detail": "low"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/6213_image_2.jpg",
            "detail": "low"
 

### Split the data into train, val and test set in the ratio of 8:1:1

In [ ]:
import json
import random
from pathlib import Path

input_file = "/content/drive/My Drive/master_thesis_project/uzh_data_for_gpt4o_description_only_detail_low.jsonl"
output_dir = "/content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only"
Path(output_dir).mkdir(parents=True, exist_ok=True)

# Load all examples
with open(input_file, "r", encoding="utf-8") as f:
    lines = f.readlines()

total = len(lines)
print(f"Total examples: {total}")

# Shuffle the data
random.shuffle(lines)

# Compute split indices
train_end_idx = int(total * 0.8)
val_end_idx = int(total * 0.9)

train = lines[:train_end_idx]
val = lines[train_end_idx:val_end_idx]
test = lines[val_end_idx:]

# Save splits
splits = {"train": train, "validation": val, "test": test}

for split_name, split_data in splits.items():
    output_file = Path(output_dir) / f"{split_name}.jsonl"
    with open(output_file, "w", encoding="utf-8") as f_out:
        for line in split_data:
            f_out.write(line)
    print(f"{split_name}: {len(split_data)} examples saved to {output_file}")

# Count total images for each split file
print("\n=== Image Counts ===")
for split_name in ["train", "validation", "test"]:
    split_file = Path(output_dir) / f"{split_name}.jsonl"
    total_images = 0

    with open(split_file, "r", encoding="utf-8") as f:
        for line in f:
            data = json.loads(line)
            for message in data.get("messages", []):
                content = message.get("content")
                if isinstance(content, list):
                    for item in content:
                        if isinstance(item, dict) and item.get("type") == "image_url":
                            total_images += 1

    print(f"{split_name}: {total_images} total images")

Total examples: 495
train: 396 examples saved to /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/train.jsonl
validation: 49 examples saved to /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/validation.jsonl
test: 50 examples saved to /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/test.jsonl

=== Image Counts ===
train: 2927 total images
validation: 402 total images
test: 385 total images


In [ ]:
# make sure the encoding is utf-8 for gpt4o fine-tuning
import chardet

with open(output_file, "rb") as f:
    raw = f.read(4096)
    print(chardet.detect(raw))

{'encoding': 'utf-8', 'confidence': 0.99, 'language': ''}


### Fine-tuning

In [ ]:
!pip install --upgrade openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 948.4/948.4 kB 15.6 MB/s eta 0:00:00
  Attempting uninstall: openai
    Found existing installation: openai 1.107.0
    Uninstalling openai-1.107.0:
      Successfully uninstalled openai-1.107.0


#### load the inspect the data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
uzh_data_dir = "/content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only"

In [ ]:
!wc -l "{uzh_data_dir}/train.jsonl"
!wc -l "{uzh_data_dir}/validation.jsonl"
!wc -l "{uzh_data_dir}/test.jsonl"

396 /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/train.jsonl
49 /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/validation.jsonl
50 /content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only/test.jsonl


In [ ]:
!head -n 10 "{uzh_data_dir}/train.jsonl"

{"messages": [{"role": "system", "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."}, {"role": "user", "content": [{"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_0.jpg", "detail": "low"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_1.jpg", "detail": "low"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_2.jpg", "detail": "low"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_3.jpg", "detail": "low"}}, {"type": "image_url", "image_url": {"url": "https://raw.githubusercontent.com/juanit

50 examples, 33717 tokens.  
https://platform.openai.com/tokenizer  

33717/50*396 = 267038.64 tokens for train.jsonl

In [ ]:
import json

# Show one pretty-printed sample from your JSONL file
with open(f"{uzh_data_dir}/train.jsonl", "r", encoding="utf-8") as f:
    for i, line in enumerate(f):
        if i == 0:  # just take the first artefact
            obj = json.loads(line)
            print(json.dumps(obj, indent=2, ensure_ascii=False))
            break

{
  "messages": [
    {
      "role": "system",
      "content": "Du bist eine Museumskuratorin, die präzise und fachlich korrekte Beschreibungen archäologischer Artefakte auf Deutsch erstellt. Verwende einen sachlichen, musealen Stil wie in Ausstellungskatalogen."
    },
    {
      "role": "user",
      "content": [
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_0.jpg",
            "detail": "low"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_1.jpg",
            "detail": "low"
          }
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/7805_image_2.jpg",
            "detail": "low"
 

#### run gpt4o fine-tuning

use the fine-tunable vision model gpt-4o-2024-08-06  
https://platform.openai.com/docs/guides/vision-fine-tuning . But in this link, it says that "Each example can have at most 10 images."

Accoding to https://learn.microsoft.com/en-us/azure/ai-foundry/openai/how-to/fine-tuning-vision :
Vision fine-tuning is supported for gpt-4o version 2024-08-06 and gpt-4.1 version 2025-04-14 models only. In this link, it says "Each example can have at most 64 images." This is written in 2025. but in this link https://platform.openai.com/docs/guides/supervised-fine-tuning gpt-4.1 version 2025-04-14 models are not listed in the vision fine-tuning page.

In [ ]:
# initiate openai client

from openai import OpenAI

client = OpenAI(api_key="")

upload the train and validation files

In [ ]:
training_file_upload_response = client.files.create(
    file=open(f"{uzh_data_dir}/train.jsonl", "rb"),
    purpose="fine-tune"
)

validation_file_upload_response = client.files.create(
    file=open(f"{uzh_data_dir}/validation.jsonl", "rb"),
    purpose="fine-tune"
)

print("training_file_upload_response:", training_file_upload_response)
print("validation_file_upload_response:", validation_file_upload_response)

training_file_upload_response: FileObject(id='file-SLRmZ9wToz56zzqsVYdrYH', bytes=901807, created_at=1761212944, filename='train.jsonl', object='file', purpose='fine-tune', status='processed', expires_at=None, status_details=None)
validation_file_upload_response: FileObject(id='file-LP4u2CFxco791qVP5PPjj9', bytes=128007, created_at=1761212945, filename='validation.jsonl', object='file', purpose='fine-tune', status='processed', expires_at=None, status_details=None)


create a fine-tuned mode

In [ ]:
from datetime import datetime
import re

def generate_finetune_suffix(project_name: str, dataset_name: str, version: str = None) -> str:
    """
    Generates a kebab-case suffix for fine-tuning jobs, where spaces are replaced with hyphens and all letters are lower-case.

    Args:
        project_name (str): Short name for the project, e.g., 'artefacts'.
        dataset_name (str): Name of the dataset, e.g., 'uzh'.
        version (str, optional): Version or date, e.g., 'v1' or '2025-10-15'. Defaults to today.

    Returns:
        str: kebab-case suffix, e.g., 'artefacts-uzh-2025-10-15'
    """
    def to_kebab(text: str) -> str:
        return re.sub(r'\s+', '-', text.strip().lower())

    if version is None:
        version = datetime.today().strftime("%Y-%m-%d")

    parts = [project_name, dataset_name, version]

    return '-'.join(to_kebab(part) for part in parts)


fine_tuning_response = client.fine_tuning.jobs.create(
    training_file=training_file_upload_response.id,
    validation_file=validation_file_upload_response.id,
    suffix=generate_finetune_suffix('artefacts', 'uzh', 'v4'),
    model="gpt-4o-2024-08-06"

)

fine_tuning_response

FineTuningJob(id='ftjob-I8bQO3Im2qZ3kWJXuhFWGf8B', created_at=1761213141, error=Error(code=None, message=None, param=None), fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(batch_size='auto', learning_rate_multiplier='auto', n_epochs='auto'), model='gpt-4o-2024-08-06', object='fine_tuning.job', organization_id='org-QVEfkKafjj7yW8YeqB6EfSev', result_files=[], seed=591655118, status='validating_files', trained_tokens=None, training_file='file-SLRmZ9wToz56zzqsVYdrYH', validation_file='file-LP4u2CFxco791qVP5PPjj9', estimated_finish=None, integrations=[], metadata=None, method=Method(type='supervised', dpo=None, reinforcement=None, supervised=SupervisedMethod(hyperparameters=SupervisedHyperparameters(batch_size='auto', learning_rate_multiplier='auto', n_epochs='auto'))), user_provided_suffix='artefacts-uzh-v4', usage_metrics=None, shared_with_openai=False, eval_id=None)

check training job status

In [ ]:
status_response = client.fine_tuning.jobs.retrieve(fine_tuning_response.id)

status_response

FineTuningJob(id='ftjob-I8bQO3Im2qZ3kWJXuhFWGf8B', created_at=1761213141, error=Error(code=None, message=None, param=None), fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(batch_size=1, learning_rate_multiplier=2.0, n_epochs=3), model='gpt-4o-2024-08-06', object='fine_tuning.job', organization_id='org-QVEfkKafjj7yW8YeqB6EfSev', result_files=[], seed=591655118, status='validating_files', trained_tokens=None, training_file='file-SLRmZ9wToz56zzqsVYdrYH', validation_file='file-LP4u2CFxco791qVP5PPjj9', estimated_finish=None, integrations=[], metadata=None, method=Method(type='supervised', dpo=None, reinforcement=None, supervised=SupervisedMethod(hyperparameters=SupervisedHyperparameters(batch_size=1, learning_rate_multiplier=2.0, n_epochs=3))), user_provided_suffix='artefacts-uzh-v4', usage_metrics=None, shared_with_openai=False, eval_id=None)

In [ ]:
from openai import OpenAI

client = OpenAI(api_key="")
status = client.fine_tuning.jobs.retrieve('ftjob-I8bQO3Im2qZ3kWJXuhFWGf8B')
status.status

'succeeded'

run inference using fine-tuned model

In [ ]:
from openai import OpenAI

uzh_data_dir = "/content/drive/My Drive/master_thesis_project/gpt4o/uzh_split_data/description_only"

client = OpenAI(api_key="")
status_response = client.fine_tuning.jobs.retrieve('ftjob-I8bQO3Im2qZ3kWJXuhFWGf8B')
fine_tuned_model = status_response.fine_tuned_model

In [ ]:
fine_tuned_model

'ft:gpt-4o-2024-08-06:university-of-zurich-department-of-history:artefacts-uzh-v4:CTnZHaC9'

In [ ]:
import json
from openai import OpenAI
import re

# Load JSONL file
def load_jsonl(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        return [json.loads(line) for line in f]

test_data = load_jsonl(f"{uzh_data_dir}/test.jsonl")

completion = client.chat.completions.create(
    model=status_response.fine_tuned_model,
    messages=test_data[0]['messages'][:-1]
)

completion.choices[0].message

ChatCompletionMessage(content='Die Mündungs- und die Halsoberseite sind tongrundig. Am Schulteransatz Strichmuster. Auf der Schulter Veloutte.\n\nOstgriechischer Aryballos.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None)

In [ ]:
completion = client.chat.completions.create(
    model=status_response.fine_tuned_model,
    messages=test_data[1]['messages'][:-1]
)

completion.choices[0].message

ChatCompletionMessage(content='Niedriger ein wenig unregelmässiger Napf mit horizontal gelochten Henkeln. Die breite untere Zone endet gegen die Basis in zwei schmalen Zonen, die jeweils durch einen umlaufenden Reifen abgetrennt sind. Danach folgt unmittelbar der Standring. Ausserhalb der Zonen sind die Henkel mit Streifen verziert. Auf Wandung und Schulter alternierende breitere und schmalere vertikale Barbotinewellen, flankiert von je einem dünnen Weinrankenstiel. Auf dem Fuss eine im Pressverfahren hergestellte Meduse.\n\nFeiner, schlammfarbener, wenig hart gebrannter Ton mit gleichmässig dünnem, glänzend braunschwarzem bis braunrotem Firnis / Glanzton. Barbotine in Form von wenig dünnflüssiger Auflagemasse.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None)

In [ ]:
import json
from openai import OpenAI
from tqdm import tqdm

# Load JSONL file
def load_jsonl(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        return [json.loads(line) for line in f]

def generate_description_only(data, model, max_samples=None):
    predictions = []

    total_samples = len(data) if max_samples is None else min(max_samples, len(data))

    # Simple progress bar
    for i, sample in enumerate(tqdm(data, total=total_samples, desc="Generating descriptions")):
        if max_samples and i >= max_samples:
            break

        messages = sample["messages"]
        ground_description = messages[2]["content"]

        # First image URL
        first_image_url = None
        for c in messages[1]["content"]:
            if c.get("type") == "image_url":
                first_image_url = c["image_url"]["url"]
                break

        # Single turn: Predict description only
        conversation = [messages[0], messages[1]]

        completion = client.chat.completions.create(
            model=model,
            messages=conversation
        )
        prediction_description = completion.choices[0].message.content.strip()

        # Save predictions
        predictions.append({
            "sample_id": i + 1,
            "first_image_url": first_image_url,
            "ground_description": ground_description,
            "prediction_description": prediction_description
        })

    return predictions

In [ ]:
# Load test set
test_data = load_jsonl(f"{uzh_data_dir}/test.jsonl")

# Generate predictions
test_preds = generate_description_only(test_data, fine_tuned_model)

# Print nicely
def print_predictions(preds):
    for r in preds:
        print(json.dumps(r, indent=2, ensure_ascii=False))

print("\n=== Test Set Description Predictions ===")
print_predictions(test_preds)

Generating descriptions: 100%|██████████| 50/50 [06:32<00:00,  7.85s/it]


=== Test Set Description Predictions ===
{
  "sample_id": 1,
  "first_image_url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/4769_image_0.jpg",
  "ground_description": "Am Hals Stabmuster. Zwischen zwei Volutenranken Vogel mit ausgebreitetem Flügel.\n\nFeiner orangebrauner hart gebrannter Ton, leicht metallisch glänzender grauschwarzer Firnis / Glanzton.",
  "prediction_description": "Hals mit Ausnahme eines tongrundigen Reifen am Ansatz gefirnisst (Glanzton). Am Schulterknick ein gelber Reifen. Eingeritzte Goachedekoration: Fünfzackiger Stern im Zentrum eingefasst von konzentrischen Kreisen.  \n\nFeiner hellbrauner hart gebrannter Ton mit glänzendem schwarzem wenig sorgfältig aufgetragenem Firnis / Glanzton."
}
{
  "sample_id": 2,
  "first_image_url": "https://raw.githubusercontent.com/juanitalfr5-pixel/uzh_artefacts_images/main/3850_image_0.jpg",
  "ground_description": "Innen bis auf einen tongrundigen Reifen auf dem Rand gefirnisst (Glanzton). I

In [ ]:
!pip install evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 2.8 MB/s eta 0:00:00


In [ ]:
!pip install rouge_score

  Preparing metadata (setup.py) ... done
  Created wheel for rouge_score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=ec390938ef968810bcd1bb74861146b4705a159836883843ab286d3ccfd714d0
  Stored in directory: /root/.cache/pip/wheels/85/9d/af/01feefbe7d55ef5468796f0c68225b6788e85d9d0a281e7a70
Successfully built rouge_score


In [ ]:
!pip install bert_score

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 2.1 MB/s eta 0:00:00


In [ ]:
import evaluate
import re

# Load metrics
bleu = evaluate.load("bleu")
rouge = evaluate.load("rouge")
meteor = evaluate.load("meteor")
bertscore = evaluate.load("bertscore")

def compute_description_metrics(predictions):
    refs = [p['ground_description'] for p in predictions]
    hyps = [p['prediction_description'] for p in predictions]

    print(f"Evaluating {len(predictions)} samples...")

    # BLEU - FIXED: Use string inputs, not tokenized
    try:
        bleu_score = bleu.compute(predictions=hyps, references=[[ref] for ref in refs])
        print(f"BLEU: {bleu_score['bleu']:.4f}")
    except Exception as e:
        print(f"BLEU failed: {e}")

    # ROUGE (working)
    try:
        rouge_score = rouge.compute(predictions=hyps, references=refs)
        print("ROUGE:", {k: round(v, 4) for k, v in rouge_score.items()})
    except Exception as e:
        print(f"ROUGE failed: {e}")

    # METEOR (working)
    try:
        meteor_score = meteor.compute(predictions=hyps, references=refs)
        print(f"METEOR: {meteor_score['meteor']:.4f}")
    except Exception as e:
        print(f"METEOR failed: {e}")

    try:
        bertscore_score = bertscore.compute(
            predictions=hyps,
            references=refs,
            lang="de"
        )
        avg_f1 = sum(bertscore_score['f1']) / len(bertscore_score['f1'])
        print(f"BERTScore F1: {avg_f1:.4f}")
    except Exception as e:
        print(f"BERTScore failed: {e}")

# Run metrics
compute_description_metrics(test_preds)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


Evaluating 50 samples...
BLEU: 0.0450
ROUGE: {'rouge1': np.float64(0.2726), 'rouge2': np.float64(0.0948), 'rougeL': np.float64(0.2054), 'rougeLsum': np.float64(0.2295)}
METEOR: 0.2782


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/714M [00:00<?, ?B/s]

BERTScore F1: 0.7239


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def compute_tfidf_metrics(predictions):
    """TF-IDF based metrics that work better for creative text"""
    refs = [p['ground_description'] for p in predictions]
    hyps = [p['prediction_description'] for p in predictions]

    # TF-IDF Vectorizer with German-specific settings
    vectorizer = TfidfVectorizer(
        max_features=1000,
        min_df=1,  # Include rare archaeological terms
        max_df=0.8,  # Exclude very common words
        stop_words=None,  # Keep all German words
        ngram_range=(1, 2)  # Include bigrams for technical terms
    )

    # Fit on all text
    all_text = refs + hyps
    vectorizer.fit(all_text)

    # Transform references and predictions
    refs_vec = vectorizer.transform(refs)
    hyps_vec = vectorizer.transform(hyps)

    # Compute cosine similarities
    cosine_sims = []
    for i in range(len(refs)):
        sim = cosine_similarity(refs_vec[i], hyps_vec[i])[0][0]
        cosine_sims.append(sim)

    # Additional metrics
    length_ratios = [len(hyp.split()) / len(ref.split()) for hyp, ref in zip(hyps, refs)]

    print("=== TF-IDF METRICS ===")
    print(f"Average Cosine Similarity: {np.mean(cosine_sims):.4f}")
    print(f"Cosine Similarity Std: {np.std(cosine_sims):.4f}")
    print(f"Average Length Ratio: {np.mean(length_ratios):.4f}")
    print(f"Length Ratio Std: {np.std(length_ratios):.4f}")

    # Quality assessment
    good_similarity = sum(1 for sim in cosine_sims if sim > 0.3) / len(cosine_sims)
    good_length = sum(1 for ratio in length_ratios if 0.5 <= ratio <= 2.0) / len(length_ratios)

    print(f"Examples with good similarity (>0.3): {good_similarity:.1%}")
    print(f"Examples with reasonable length (0.5-2.0x): {good_length:.1%}")

    return {
        'cosine_similarities': cosine_sims,
        'length_ratios': length_ratios,
        'avg_cosine': np.mean(cosine_sims),
        'avg_length_ratio': np.mean(length_ratios)
    }

tfidf_high = compute_tfidf_metrics(test_preds)

=== TF-IDF METRICS ===
Average Cosine Similarity: 0.1879
Cosine Similarity Std: 0.1251
Average Length Ratio: 1.9318
Length Ratio Std: 3.8159
Examples with good similarity (>0.3): 10.0%
Examples with reasonable length (0.5-2.0x): 60.0%
